### PR TITLE
Implement Iterator.prototype.take

### LINKED ISSUE #80: Add `Iterator.prototype.take` to the adventure pack
See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/take

Probably worth doing at the same time as #81.

### PR DESCRIPTION
**Pull request by simona1**
Closes #80

### CODE REVIEW

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    const result = iteratorToArray([1, 2, 3, 4, 5].values().take(4));
+    expect(result).toEqual([1, 2, 3, 4]);
+  });
+
+  it("takes all elements if the specified number is greater than the iterator length", () => {
+    const result = iteratorToArray([1, 2, 3].values().take(5));
+    expect(result).toEqual([1, 2, 3]);
+  });
+
+  it("returns an empty array if the specified number is 0", () => {
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:23]:
```suggestion
  it("returns an empty iterator if the specified number is 0", () => {
```

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    const result = iteratorToArray([1, 2, 3, 4, 5].values().take(0));
+    expect(result).toEqual([]);
+  });
+
+  it("returns an empty array for an empty iterator", () => {
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:28]:
```suggestion
  it("returns an empty iterator when taking from an empty iterator", () => {
```

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      let i = -1;
+      while (true) {
+        yield ++i;
+      }
+    }
+    const result = iteratorToArray(infiniteNumbers().take(5));
+    expect(result).toEqual([0, 1, 2, 3, 4]);
+  });
+
+  it("throws an error when taking 5 elements from an infinite iterator that errors", () => {
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:44]: It's not actually an infinite iterator -- it's just an iterator that throws! Please update the variable and test name accordingly.

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      yield 3;
+      throw new Error();
+    }
+
+    expect(() => {
+      iteratorToArray(infiniteNumbers().take(5));
+    }).toThrow();
+  });
+
+  it("does not throw an error when taking 1 element from an infinite iterator that errors later", () => {
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:55]:
```suggestion
  it("ignores errors in the underlying iterator that happen past the limit of elements taken", () => {
```

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    function* infiniteNumbers() {
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:56]: Please update name to something more appropriate.

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      yield 3;
+      throw new Error();
+    }
+
+    expect(() => {
+      iteratorToArray(infiniteNumbers().take(1));
+    }).not.toThrow();
+  });
+
+  it("throws a RangeError if count is NaN", () => {
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:66]: All of the `RangeError` tests can probably be shortened using an `it.each`. See https://jestjs.io/docs/api#testeachtablename-fn-timeout

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import "../Iterator.prototype.toIterable";
+import { iteratorPrototype } from "../Iterator.prototype";
+
+declare global {
+  interface Iterator<T> {
+    take(this: Iterator<T>, count: number): IterableIterator<T>;
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.ts:6]: Let's call it limit, like MDN does: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/take#limit

(here and below)

**Review by miorel — CHANGES_REQUESTED** [review]
Thanks for working on this! Added some inline comments.

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
+import { describe, expect, it } from "@jest/globals";
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:1]: Btw, feel free to update the contributors list in the Adventure Pack's `package.json` to include your name! (i.e. the file `workspaces/adventure-pack/package.json`)

File: workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+import { iteratorPrototype } from "../Iterator.prototype";
+delete (iteratorPrototype as unknown as Record<string, unknown>).take;
+
+import "./index";
+
+function iteratorToArray<T>(iterator: IterableIterator<T>): T[] {
+  return [...iterator];
+}
```
Comment (miorel) [inline workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:10]: Does it work if we replace the callsites of this function with `Array.from`?

Comment (simona1) [reply workspaces/adventure-pack/goodies/typescript/Iterator.prototype.take/index.test.ts:10]: it does

**Review by elimanzo — APPROVED** [review]
Nice I never learned about `take` until today. Thank you!

**Pull request merged**